gcs/smart_entry.hh

```diff
@@ -0,0 +1,130 @@
+#ifndef GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_SMART_ENTRY_HH
+#define GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_SMART_ENTRY_HH
+
+#include <variant>
+#include <vector>
+#include <gcs/variable_id.hh>
+#include <gcs/integer.hh>
+#include <algorithm> //std::sort
+using namespace gcs;
+
+enum ConstraintType {
+    LESS_THAN,
+    LESS_THAN_EQUAL,
+    EQUAL,
+    NOT_EQUAL,
+    GREATER_THAN,
+    GREATER_THAN_EQUAL,
+    IN,
+    NOT_IN
+};
+
+/**
+ * \brief Smart entries for smart tables
+ */
+struct BinaryEntry {
+    IntegerVariableID var_1;
+    IntegerVariableID var_2;
+    ConstraintType constraint_type;
+
+    // Constructor
+    BinaryEntry(IntegerVariableID var_1, IntegerVariableID var_2, ConstraintType constraint_type)
+    : var_1(var_1), var_2(var_2), constraint_type(constraint_type) {};
+};
+
+struct UnaryValueEntry {
+    IntegerVariableID var;
+    Integer value;
+    ConstraintType constraint_type;
+
+    // Constructor
+    UnaryValueEntry(const IntegerVariableID &var, const Integer &value, ConstraintType constraint_type)
+    : var(var), value(value),constraint_type(constraint_type) {};
+};
+
+struct UnarySetEntry {
+    IntegerVariableID var;
+    std::vector<Integer> values;
+    ConstraintType constraint_type;
+
+    // Constructor
+    UnarySetEntry(const IntegerVariableID &var, const std::vector<Integer> &values, ConstraintType constraint_type)
```

Comment:
Again, no ctor needed, but if you did need one, I think it's best to pass the vector by value and then use a move.